src/snowflake/snowpark/dataframe.py

```diff
@@ -1650,18 +1650,18 @@ def _union_by_name_internal(
         else:
             right_child = self._with_plan(Project(names, other._plan))
 
-        union_plan = UnionPlan(self._plan, right_child._plan, is_all)
         if self._session.sql_simplifier_enabled:
             df = self._with_plan(
-                SelectStatement(
-                    from_=SelectSnowflakePlan(
-                        snowflake_plan=union_plan, analyzer=self._session._analyzer
+                self._select_statement.set_operator(
+                    right_child._select_statement
+                    or SelectSnowflakePlan(
+                        right_child._plan, analyzer=self._session._analyzer
                     ),
-                    analyzer=self._session._analyzer,
+                    operator=SET_UNION,
```

Comment:
This needs to be `SET_UNION_ALL` when `is_all` is True.

Reply:
yep. I just realized this error too. Good to know that SQL simplifier tests are working as expected